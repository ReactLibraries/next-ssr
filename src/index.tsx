"use client";

import React, {
  ReactNode,
  useContext,
  useId,
  useRef,
  useCallback,
  useSyncExternalStore,
  createContext,
  useEffect,
  FC,
} from "react";

const DATA_NAME = "__NEXT_DATA_PROMISE__";
const isServerSide = typeof window === "undefined";

type StateType<T> = {
  data?: T;
  error?: unknown;
  isLoading: boolean;
  fetcher: () => Promise<T>;
  promise?: Promise<T>;
};
type Render = () => void;
type ContextType = {
  values: { [key: string]: StateType<unknown> };
  promises: Set<Promise<unknown>>;
  resolve: () => void;
  finished: boolean;
  promise: Promise<void>;
  renderMap: Map<string | number, Set<Render>>;
  isDataRender?: boolean;
};

/**
 * Context for asynchronous data management
 */
const promiseContext = createContext<ContextType>(undefined as never);

/**
 * Rendering event propagation
 */
const render = (
  renderMap: Map<string | number, Set<Render>>,
  key: string | number
) => renderMap.get(key)?.forEach((render) => render());

/**
 * Asynchronous data loading
 */
const loader = <T,>(
  key: string | number,
  context: ContextType,
  fetcher?: () => Promise<T>
) => {
  const { promises, values, renderMap } = context;
  const _fetcher = fetcher ?? values[key]?.fetcher;
  if (!_fetcher) throw new Error("Empty by fetcher");
  const value: StateType<T> = {
    data: values[key]?.data as T,
    error: undefined,
    isLoading: true,
    fetcher: _fetcher as () => Promise<T>,
  };
  values[key] = value;
  render(renderMap, key);
  const promise = _fetcher();
  value.promise = promise as Promise<T>;
  if (isServerSide) {
    promises.add(promise);
  }
  promise
    .then((v) => {
      values[key] = {
        data: v,
        error: undefined,
        isLoading: false,
        fetcher: _fetcher,
      };
      render(renderMap, key);
    })
    .catch((error) => {
      values[key] = {
        data: undefined,
        error,
        isLoading: false,
        fetcher: _fetcher,
      };
      render(renderMap, key);
    });
  return promise;
};

/**
 * hook for re-loading
 */
export const useReload = (key: string | number) => {
  const context = useContext(promiseContext);
  return useCallback(() => {
    loader(key, context);
  }, [context, key]);
};

/**
 * Asynchronous data acquisition hook for SSR
 */
export const useSSR = <T,>(
  fetcher: () => Promise<T>,
  { key }: { key?: string | number } = {}
): StateType<T> & { reload: () => void } => {
  const context = useContext(promiseContext);
  const { values, renderMap } = context;
  const id = useId();
  const cacheKey = key ?? id;

  const value = useSyncExternalStore(
    (callback) => {
      const renderSet = renderMap.get(cacheKey) ?? new Set<Render>();
      renderMap.set(cacheKey, renderSet);
      renderSet.add(callback);
      return () => renderSet.delete(callback);
    },
    () => values[cacheKey] as StateType<T>,
    () => values[cacheKey] as StateType<T>
  );

  const reload = useCallback(() => {
    return loader(cacheKey, context, fetcher);
  }, [cacheKey, context, fetcher]);
  if (!value) {
    const promise = reload();
    if (isServerSide) {
      throw promise;
    }
  } else if (value.isLoading) {
    if (isServerSide) throw value.promise;
  } else if (!value.fetcher) {
    value.fetcher = fetcher;
  }

  return { ...value, reload };
};

/**
 * Transfer of SSR data to clients
 */
const DataRender = () => {
  const ssrContext = useContext(promiseContext);
  if (isServerSide && !ssrContext.finished) {
    let length: number;
    while ((length = ssrContext.promises.size)) {
      throw Promise.allSettled(ssrContext.promises).then(() => {
        if (length === ssrContext.promises.size) {
          ssrContext.promises.clear();
        }
      });
    }
  }
  ssrContext.finished = true;
  ssrContext.resolve();
  return (
    <script
      id={DATA_NAME}
      type="application/json"
      dangerouslySetInnerHTML={{
        __html: JSON.stringify(ssrContext.values).replace(/</g, "\\u003c"),
      }}
    />
  );
};
export const SSRDataRender = () => {
  const ssrContext = useContext(promiseContext);
  if (ssrContext.isDataRender) return null;
  ssrContext.isDataRender = true;
  return <DataRender />;
};
/**
 * Context data initialization hook
 */
const useContextValue = () => {
  let resolve: () => void = () => {};
  const promise = new Promise<void>((r) => {
    resolve = r;
  });
  const refContext = useRef<ContextType>({
    values: {},
    promises: new Set(),
    finished: false,
    renderMap: new Map<string, Set<Render>>(),
    resolve,
    promise,
  });
  if (!isServerSide && !refContext.current.finished) {
    const node = document.getElementById(DATA_NAME);
    if (node) refContext.current.values = JSON.parse(node.innerHTML);
    refContext.current.finished = true;
  }
  return refContext.current;
};

/**
 * Provider for asynchronous data management
 */
export const SSRProvider = ({ children }: { children: ReactNode }) => {
  const value = useContextValue();
  return (
    <SSRHeadProvider>
      <promiseContext.Provider value={value}>
        {children}
        <SSRDataRender />
      </promiseContext.Provider>
    </SSRHeadProvider>
  );
};

export const SSRWait = ({ children }: { children: ReactNode }) => {
  const ssrContext = useContext(promiseContext);
  if (isServerSide) {
    if (!ssrContext.finished) {
      throw ssrContext.promise;
    }
  }
  return children;
};

//---------------------

const HEAD_DATA_NAME = "__REMIX_HEAD_VALUE__";

export type HeadContextType<T = ReactNode[]> = {
  state: T;
  storeChanges: Set<() => void>;
  dispatch: (callback: (state: T) => T) => void;
  subscribe: (onStoreChange: () => void) => () => void;
  promise: Promise<void>;
  resolve: () => void;
  finished: boolean;
};

export const useCreateHeadContext = <T,>(initState: () => T) => {
  let resolve: () => void = () => {};
  const promise = new Promise<void>((r) => {
    resolve = r;
  });
  const context = useRef<HeadContextType<T>>({
    state: initState(),
    storeChanges: new Set(),
    dispatch: (callback) => {
      context.state = callback(context.state);
      context.storeChanges.forEach((storeChange) => storeChange());
    },
    subscribe: (onStoreChange) => {
      context.storeChanges.add(onStoreChange);
      return () => {
        context.storeChanges.delete(onStoreChange);
      };
    },
    promise,
    resolve,
    finished: false,
  }).current;
  return context;
};

const HeadContext = createContext<
  HeadContextType<{ type: string; props: Record<string, unknown> }[][]>
>(undefined as never);

const Wait = () => {
  const context = useContext(HeadContext);
  if (!context.finished) {
    context.finished = true;
    context.resolve();
  }
  return null;
};

const SSRHeadProvider = ({ children }: { children: ReactNode }) => {
  const context = useCreateHeadContext<
    { type: string; props: Record<string, unknown> }[][]
  >(() => {
    if (typeof window !== "undefined") {
      return [
        JSON.parse(
          document.querySelector(`script#${HEAD_DATA_NAME}`)?.textContent ??
            "{}"
        ),
      ];
    }
    return [[]];
  });
  return (
    <HeadContext.Provider value={context}>
      {children}
      <Wait />
    </HeadContext.Provider>
  );
};

export const SSRHeadRoot: FC = () => {
  return (
    <SSRWait>
      <HeadRoot />
    </SSRWait>
  );
};

const HeadRoot: FC = () => {
  const context = useContext(HeadContext);
  if (isServerSide && !context.finished) {
    throw context.promise;
  }
  const state = useSyncExternalStore(
    context.subscribe,
    () => context.state,
    () => context.state
  );
  useEffect(() => {
    context.dispatch(() => {
      return [];
    });
  }, [context]);
  const heads = state.flat();
  return (
    <>
      <script
        id={HEAD_DATA_NAME}
        type="application/json"
        dangerouslySetInnerHTML={{
          __html: JSON.stringify(heads).replace(/</g, "\\u003c"),
        }}
      />
      {heads.map(({ type: Tag, props }, index) => (
        <Tag key={`HEAD${Tag}${index}`} {...props} />
      ))}
    </>
  );
};

export const SSRHead: FC<{ children: ReactNode }> = ({ children }) => {
  const context = useContext(HeadContext);
  useEffect(() => {
    const value = extractInfoFromChildren(children);
    context.dispatch((heads) => [...heads, value]);
    return () => {
      context.dispatch((heads) => heads.filter((head) => head !== value));
    };
  }, [children, context]);

  if (isServerSide) {
    context.dispatch((heads) => [...heads, extractInfoFromChildren(children)]);
  }
  return null;
};

const extractInfoFromChildren = (
  children: ReactNode
): { type: string; props: Record<string, unknown> }[] =>
  React.Children.toArray(children).flatMap((child) => {
    if (React.isValidElement(child)) {
      if (child.type === React.Fragment) {
        return extractInfoFromChildren(child.props.children);
      }
      if (typeof child.type === "string") {
        return [{ type: child.type, props: child.props }];
      }
    }
    return [];
  });
