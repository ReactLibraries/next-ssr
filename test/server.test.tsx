import { EventEmitter } from 'events';
import { JSDOM } from 'jsdom';
import React, { ReactNode } from 'react';
import { renderToPipeableStream } from 'react-dom/server';
import { SSRProvider, useSSR } from '../src';

class StreamReader extends EventEmitter {
  writable = true;
  private chunks: Buffer[] = [];
  constructor(private cb: (s: string) => void) {
    super();
  }
  write(chunk: Uint8Array | string) {
    this.chunks.push(Buffer.from(chunk));
    return true;
  }
  end() {
    this.cb(Buffer.concat(this.chunks).toString('utf8'));
    return this;
  }
}

let prev;
beforeEach(() => {
  prev = global.IS_REACT_ACT_ENVIRONMENT;
  global.IS_REACT_ACT_ENVIRONMENT = true;
});

afterEach(() => {
  global.IS_REACT_ACT_ENVIRONMENT = prev;
});

const Component01 = () => {
  const { data } = useSSR<number>(async () => {
    return 100;
  });
  return <div id="data1">{data}</div>;
};

const Component02 = () => {
  const { data } = useSSR<number>(
    async () => {
      return 200;
    },
    { key: 'Component02' }
  );
  return <div id="data2">{data}</div>;
};

const Component03 = () => {
  const { error } = useSSR<number>(
    async () => {
      throw 'error';
    },
    { key: 'Component03' }
  );
  return <div id="data3">{String(error)}</div>;
};

const Component04 = () => {
  const { data } = useSSR<number>(async () => 100, { key: 'Component04' });
  return <div id="data4">{data}</div>;
};

const render = (element: ReactNode) =>
  new Promise<string>((resolve, reject) => {
    const stream = renderToPipeableStream(element, {
      onError(error) {
        reject(error);
      },
      onAllReady() {
        stream.pipe(
          new StreamReader((html) => {
            resolve(html);
          })
        );
      },
    });
  });

it('SSR', async () => {
  const stream = await render(
    <SSRProvider>
      <Component01 />
      <Component02 />
      <Component03 />
    </SSRProvider>
  );
  const document = new JSDOM(stream).window.document;
  expect(document.getElementById('data1')?.innerHTML).toBe('100');
  expect(document.getElementById('data2')?.innerHTML).toBe('200');
  expect(document.getElementById('data3')?.innerHTML).toBe('error');
  expect(stream).toMatchSnapshot();
});

it('SSR2', async () => {
  const stream = await render(
    <SSRProvider>
      <Component04 />
      <Component04 />
      <Component04 />
    </SSRProvider>
  );
  const document = new JSDOM(stream).window.document;
  expect(document.getElementById('data4')?.innerHTML).toBe('100');
  expect(stream).toMatchSnapshot();
});
