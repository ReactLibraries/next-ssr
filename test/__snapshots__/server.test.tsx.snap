// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSR 1`] = `"<div id="data1">100</div><div id="data2">200</div><div id="data3">error</div><script id="__NEXT_DATA_PROMISE__" type="application/json">{":R5:":{"data":100,"isLoading":false},"Component02":{"data":200,"isLoading":false},"Component03":{"error":"error","isLoading":false}}</script>"`;

exports[`SSR2 1`] = `"<div id="data4">100</div><div id="data4">100</div><div id="data4">100</div><script id="__NEXT_DATA_PROMISE__" type="application/json">{"Component04":{"data":100,"isLoading":false}}</script>"`;
